src/runtime/propensity-server.js

```diff
@@ -0,0 +1,262 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {parseJson} from '../utils/json';
+export class XhrInterface {
+
+  /**
+   * Map of error code to error message
+   * @param {!Object.<number, string>} errorMap
+   */
+  constructor(errorMap) {
+    this.errorMap_ = errorMap;
+  }
+
+  /**
+   * Send request
+   * @param {string} url
+   * @param {!../utils/xhr.FetchInitDef} init
+   * @return {!Promise<!Response>}
+   */
+  sendRequest(url, init) {
+    return new Promise((resolve, reject) => {
+      /** @type {XMLHttpRequest} */
+      const xhr = new XMLHttpRequest();
+      const method = init.method || 'GET';
+      xhr.open(method, url, true);
+      if (init.headers) {
+        Object.keys(init.headers).forEach(function(header) {
+          xhr.setRequestHeader(header, init.headers[header]);
+        });
+      }
+      if (init.credentials) {
+        xhr.withCredentials = true;
+      }
+      xhr.onreadystatechange = () => {
+        if (xhr.readyState < /* STATUS_RECEIVED */ 2) {
+          return;
+        }
+        if (xhr.readyState == /* COMPLETE */ 4) {
+          if (xhr.status in this.errorMap_) {
+            let errorMessage = xhr.status + ':' + this.errorMap_[xhr.status];
+            if (xhr.statusText) {
+              errorMessage = errorMessage + ' ' + xhr.statusText;
+            }
+            reject(new Error(errorMessage));
+          } else if (xhr.status < 200 || xhr.status > 300) {
+            const errorMessage = xhr.status + ':' + xhr.statusText;
+            reject(new Error(errorMessage));
+          } else {
+            const response = new Response();
+            response.status = xhr.status;
+            response.ok = true;
+            response.statusText = xhr.statusText;
+            xhr.bodyUsed = true;
+            response.responseText = xhr.responseText;
+            response.url = xhr.responseURL;
+            response.headers = this.getResponseHeader_(
+                xhr.getAllResponseHeaders());
+            resolve(response);
+          }
+        }
+      };
+      xhr.onerror = () => {
+        reject(new Error('Network failure'));
+      };
+      xhr.onabort = () => {
+        reject(new Error('Request aborted'));
+      };
+      if (init.method == 'POST') {
+        xhr.send(init.body);
+      } else {
+        xhr.send();
+      }
+    });
+  }
+
+  /**
+   * Creates header from XHR header
+   * @param {?string} responseHeaders
+   * @return {!Headers}
+   */
+  getResponseHeader_(responseHeaders) {
+    const headerMap = new Headers();
+    if (!responseHeaders) {
+      return headerMap;
+    }
+    const arr = responseHeaders.trim().split(/[/r/n]+/);
+    arr.forEach(function(line) {
+      const parts = line.split(': ');
```

Comment:
nit: you could put ': ' into a const variable

Reply:
I'll be removing this piece of code anyway. So won't address this comment.